core/src/main/scala/kafka/server/ReplicaAlterLogDirsThread.scala

```diff
@@ -186,16 +187,36 @@ class ReplicaAlterLogDirsThread(name: String,
           info(s"Retrying leaderEpoch request for partition $topicPartition as the current replica reported an error: ${epochOffset.error}")
           partitionsWithError += topicPartition
         } else {
-          val fetchOffset =
-            if (epochOffset.endOffset == UNDEFINED_EPOCH_OFFSET)
-              partitionStates.stateValue(topicPartition).fetchOffset
-            else if (epochOffset.endOffset >= futureReplica.logEndOffset.messageOffset)
-              futureReplica.logEndOffset.messageOffset
-            else
-              epochOffset.endOffset
-
-          partition.truncateTo(fetchOffset, isFuture = true)
-          fetchOffsets.put(topicPartition, fetchOffset)
+          val offsetTruncationState =
+            if (epochOffset.endOffset == UNDEFINED_EPOCH_OFFSET) {
+              // since both replicas are on the same broker, they are using same version of OffsetForLeaderEpoch
+              // request/response, so we cannot get a valid epoch with undefined offset; log warn
+              // in case that happens so we know we made a wrong assumption somewhere
+              if (epochOffset.leaderEpoch != UNDEFINED_EPOCH)
+                info(s"Replica responded with undefined offset but a valid leader epoch ${epochOffset.leaderEpoch} for $topicPartition")
+              OffsetTruncationState(partitionStates.stateValue(topicPartition).fetchOffset, truncationCompleted = true)
+            } else {
+              // get (leader epoch, end offset) pair that corresponds to the largest leader epoch
+              // less than or equal to the requested epoch.
+              val (futureLeaderEpoch, futureEndOffset) = futureReplica.epochs.get.endOffsetFor (epochOffset.leaderEpoch)
+              if (futureEndOffset == UNDEFINED_EPOCH_OFFSET) {
+                // This can happen if replica was not tracking offsets at that point (before the
+                // upgrade, or if this broker is new).
+                // to be safe, we truncate to min of start offset and leader's end offset
+                val offsetToTruncateTo = min(partitionStates.stateValue(topicPartition).fetchOffset, epochOffset.endOffset)
+                finalFetchLeaderEpochOffset(offsetToTruncateTo, epochOffset.endOffset, futureReplica, isFutureReplica = true)
+              } else if (futureLeaderEpoch != epochOffset.leaderEpoch) {
+                // the replica does not know about the epoch that leader replied with
+                val intermediateOffsetToTruncateTo = min(futureEndOffset, futureReplica.logEndOffset.messageOffset)
+                OffsetTruncationState(intermediateOffsetToTruncateTo, truncationCompleted = false)
+              } else {
+                val offsetToTruncateTo = min(futureEndOffset, epochOffset.endOffset)
+                finalFetchLeaderEpochOffset(offsetToTruncateTo, epochOffset.endOffset, futureReplica, isFutureReplica = true)
+              }
+            }
```

Comment:
The above logic may still be needed in the following sequence: (1) future replica copies data above HW from current replica; (2) future replica is offline (e.g. disk failure); (3) current replica truncates data above HW and re-replicated new data from leader on the truncated offsets. To avoid duplicates, perhaps we can share the code between ReplicaFetchThread and here?